fe1-web/src/features/wallet/reducer/DigitalCashReducer.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * This error is disabled since reducers use the createSlice function, which requires the user to
+ * param-reassign. Please do not disable other errors.
+ */
+/* eslint-disable no-param-reassign */
+import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit';
+
+import { Hash } from 'core/objects';
+import { COINBASE_HASH } from 'resources/const';
+
+import { TransactionState } from '../objects/transaction';
+
+export interface DigitalCashReducerState {
+  /**
+   * A mapping between the public key hashes and their respective balances
+   */
+  balances: Record<string, number>;
+
+  /**
+   * Every hash of transactions received for this lao
+   */
+  allTransactionsHash: string[];
+
+  /**
+   * Transactions by their transaction hash (transaction id)
+   */
+  transactionsByHash: Record<string, TransactionState>;
+
+  /**
+   * A mapping between public key hashes and an array of hash of transactions which contains
+   * this public key hash in one or more of their TxOuts
+   */
+  transactionsByPubHash: Record<string, string[]>;
+}
+export interface DigitalCashLaoReducerState {
+  byLaoId: Record<string, DigitalCashReducerState>;
+}
+
+/* Initial state of the digital cash transactions */
+const initialState: DigitalCashLaoReducerState = {
+  byLaoId: {},
+};
+
+/* Name of digital cash slice in storage */
+export const DIGITAL_CASH_REDUCER_PATH = 'digitalCash';
+
+const digitalCashSlice = createSlice({
+  name: DIGITAL_CASH_REDUCER_PATH,
+  initialState,
+  reducers: {
+    /*
+     * Adds a transaction to the state
+     * We are trusting information of the transaction object, we do not verify it
+     */
+    addTransaction: (
+      state,
+      action: PayloadAction<{
+        laoId: string;
+        transactionState: TransactionState;
+      }>,
+    ) => {
+      const { laoId, transactionState } = action.payload;
+
+      if (!transactionState.transactionId) {
+        throw new Error('The transaction id of the added transaction is not defined');
+      }
+      const transactionHash = transactionState.transactionId;
+
+      /**
+       * If state is empty for given lao or roll call, we should create the initial objects
+       */
+      if (!(laoId in state.byLaoId)) {
+        state.byLaoId[laoId] = {
+          balances: {},
+          allTransactionsHash: [],
+          transactionsByHash: {},
+          transactionsByPubHash: {},
+        };
+      }
+
+      const laoState: DigitalCashReducerState = state.byLaoId[laoId];
+
+      laoState.transactionsByHash[transactionHash] = transactionState;
+      laoState.allTransactionsHash.push(transactionHash);
```

Comment:
my 2 miniLAOs: `allTransactionsHash` is equal to the set of all keys in `transactionsByHash` so it's still duplicative. Just like `allTransactions` was duplicative of the set of its values. (This could be me being bad at TS)

Reply:
Yes this is true, it `allTransactions` is equal to `Object.keys(transactionsByHash)` and could in theory be removed. But then this is a pattern used all over the codebase and in general with redux:

https://stackoverflow.com/a/61959401/2897827

Reply:
Yes, it now all depends on performance:

-Is calling Object.keys(transactionsByHash) each time much slower than having the object `allTransactionsHash`
(of course this might only be an issue when we reach a consequent number of transactions)

Order (we would use this for the transaction history) seems [to not be an issue here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys):

> The ordering of the properties is the same as that given by looping over the properties of the object manually.

This seems to mean that it would give us the order in which we added each keys.

What do you think @Xelowak @Tyratox ?

Reply:
@Tyratox Seems like I type too slowly

Reply:
yep sounds good to me